### ISSUE TITLE
Codebase/models as baseline: Performances with other losses

### ISSUE THREAD

**Issue by MauritsBleeker** [naver-ai/pcme#4]
Hi,

First of all, great work!

In my opinion, the codebase of this work can serve as a solid baseline for trying new optimization functions for image-text matching. 

I was wondering, do you have any results on the performance of other optimization functions such as the Triplet loss, Triplet loss with semi-hard negatives, or the InfoNCE loss using this codebase?
In the paper, PCME is compared with several other methods, such as VSRN, PVSE and  VSE++. However, these methods are optimized with the Triplet loss with semi-hard negatives. Do you have any insights on how much better PCME is compared to InfoNCE/Triplet loss (SH) (when optimizing with the same method and training hyper-parameters). 

Thanks,
Maurits

**Comment by SanghyukChun** [comment]
@MauritsBleeker 
Hello, thanks for reaching out to me.
Thanks for suggesting a great idea! I was planning to compare all losses before, but I couldn't make them because I was too busy in recent days. I do not have a specific comparison result as you suggested right now, but I am interested in your suggestion.

First of all, my implementation is highly flexible in terms of new objective functions. You can implement any loss function in https://github.com/naver-ai/pcme/tree/main/criterions (with a proper function input and output format) and make a config file as https://github.com/naver-ai/pcme/blob/main/config/coco/pcme_coco.yaml

Also, VSE++ is the same as `Triplet with hardest negative mining`, and VSE0 is the same as `Triplet without mining`.
I tried [distance-based miner](https://kevinmusgrave.github.io/pytorch-metric-learning/miners/#distanceweightedminer) to PCME before (haven't tried for the Triplet loss), but there was no significant improvement for PCME.

I want to note that "VSRN, PVSE and VSE++" are not optimized by semi-hard negatives, but the hardest negatives where semi-hard negatives are selected from the negatives that are not closer to the anchors than positives but have a distance less than a predefined margin. (Please see the below figure that I borrowed from https://omoindrot.github.io/triplet-loss)
<img src="https://user-images.githubusercontent.com/4919315/143371802-cd9473ef-4d22-4693-80cb-fa04be1f85f4.png" width=300px />

I presume that PCME will be better than Triplet variants with mining because PCME does not need any mining (it uses the full data points as InfoNCE or CLIP loss) and PCME can consider hidden false negatives in the dataset (There are so many false positives in the dataset as the figure below)
<img src="https://user-images.githubusercontent.com/4919315/143372265-2b36fc61-d4b5-42d6-a120-229e174b79d3.png" width=600px />

I am not confident with comparisons between PCME and InfoNCE. They are similar, but PCME is a pair-wise loss and InfoNCE is a batch-wise loss. I presume that the batch size could affect a lot to the losses, but I don't have a specific experience to compare two losses yet (but I am planning to)

**Comment by SanghyukChun** [comment]
I am planning to resume a project for the image-text matching task
(based on my new evaluation dataset: https://github.com/naver-ai/eccv-caption)
Honestly, it will take a very long time... perhaps one year. I really do not have enough time to write some codes.
However, when I release the project, I will inform you

Best,
Sanghyuk

**Comment by SanghyukChun** [comment]
@MauritsBleeker 
Hey, I remember this issue.
I don't know whether you are working in this field or not, but I did release the new codebase with multiple loss functions and different backbones (although, the main implementation is only for CLIP ViT backbones, basically, it would work for other backbones with a minor modification)

Please check this repository for more details
https://github.com/naver-ai/pcmepp

**Comment by MauritsBleeker** [comment]
Nice! Thanks a lot Sanghyuk.
